lib/edgehog_device_forwarder/termination_callbacks.ex

```diff
@@ -0,0 +1,62 @@
+# Copyright 2023 SECO Mind Srl
+# SPDX-License-Identifier: Apache-2.0
+
+defmodule EdgehogDeviceForwarder.TerminationCallbacks do
+  use GenServer
+
+  @table :termination_callbacks_table
+
+  @spec add(pid, (-> any)) :: :ok
+  def add(pid, on_close) when is_pid(pid) and is_function(on_close, 0) do
+    GenServer.cast(__MODULE__, {:add, pid, on_close})
+  end
+
+  @spec start_link(any) :: GenServer.on_start()
+  def start_link(opts \\ []) do
+    GenServer.start_link(__MODULE__, opts, name: __MODULE__)
+  end
+
+  def cache_spec do
+    params = [name: @table, ttl_check_interval: false]
+    id = {ConCache, @table}
+
+    {ConCache, params}
+    |> Supervisor.child_spec(id: id)
+  end
+
+  @impl GenServer
+  def init(_opts) do
+    # If the server is restarted, our pid changed
+    #   hence, to receive the :DOWN, we need to re-monitor existing processes
+    #   Process.monitor also sends the :DOWN for already dead pid, so we don't need to check that.
+    ConCache.ets(@table)
+    |> :ets.tab2list()
+    |> Enum.map(fn {pid, _} -> pid end)
+    |> Enum.each(&Process.monitor/1)
+
+    {:ok, nil}
+  end
+
+  @impl GenServer
+  def handle_cast({:add, pid, on_close}, state)
+      when is_pid(pid) and is_function(on_close, 0) do
+    Process.monitor(pid)
+    ConCache.put(@table, pid, on_close)
+
+    {:noreply, state}
+  end
+
+  @impl GenServer
+  def handle_info({:DOWN, _ref, :process, pid, _reason}, state) do
+    callback =
+      ConCache.isolated(@table, pid, fn ->
+        callback = ConCache.get(@table, pid)
+        ConCache.dirty_delete(@table, pid)
```

Comment:
This does not guarantee that the callback is executed before the entry is deleted from the table. 
On the other hand, executing the callback in the `ConCache.isolated` context is a no-go.


A possibility would be to split cache read from entry deletion:
1. You read the entry from cache
2. You run the callback
3. You delete the cached entry

However, this too  would not give a transactional guarantee. But, this would allow to add another process that checks if a callback for a dead process has been executed and possibly repeat it, then garbage collect dead processes removing them from the cache. I'm not sure that spinning up all of this is what we want, tough.

Another possibility is spawning a task that executes the callback in the `ConCache.isolated` context. This way you'd have the soft guarantee that the callback has at least been scheduled for execution before the related entry has been deleted:
```elixir
      ConCache.isolated(@table, pid, fn ->
        callback = ConCache.get(@table, pid)
        unless callback == nil, do: Task.start(callback/0)
        ConCache.dirty_delete(@table, pid)
      end)
```

Reply:
For our purpose, I think it's safe to use the first approach you mentioned in its simple form (ie, running the callback before deleting it)

This would guarantee that the callback is run at least one time before it gets deleted from the cache.
Right now, this works fine with the callbacks we use.

We can always complicate the structure if future problems require so.